crates/holochain_state/src/mutations.rs

```diff
@@ -0,0 +1,513 @@
+        $txn.execute(
+            &sql,
+            &[(":hash", &$hash as &dyn holochain_sqlite::rusqlite::ToSql)],
+        )
+    }};
+}
+
+/// Insert a [`DhtOp`] into the [`Scratch`].
+pub fn insert_op_scratch(scratch: &mut Scratch, op: DhtOpHashed) -> StateMutationResult<()> {
+    let (op, _) = op.into_inner();
+    let op_light = op.to_light();
+    let header = op.header();
+    let signature = op.signature().clone();
+    if let Some(entry) = op.entry() {
+        let entry_hashed = EntryHashed::with_pre_hashed(
+            entry.clone(),
+            header
+                .entry_hash()
+                .ok_or_else(|| DhtOpError::HeaderWithoutEntry(header.clone()))?
+                .clone(),
+        );
+        scratch.add_entry(entry_hashed);
+    }
+    let header_hashed = HeaderHashed::with_pre_hashed(header, op_light.header_hash().to_owned());
+    let header_hashed = SignedHeaderHashed::with_presigned(header_hashed, signature);
+    scratch.add_header(header_hashed);
+    Ok(())
+}
+
+pub fn insert_element_scratch(scratch: &mut Scratch, element: Element) {
+    let (header, entry) = element.into_inner();
+    scratch.add_header(header);
+    if let Some(entry) = entry.into_option() {
+        scratch.add_entry(EntryHashed::from_content_sync(entry))
+    }
+}
+
+/// Insert a [`DhtOp`] into the database.
+pub fn insert_op(
+    txn: &mut Transaction,
+    op: DhtOpHashed,
+    is_authored: bool,
+) -> StateMutationResult<()> {
+    let (op, hash) = op.into_inner();
+    let op_light = op.to_light();
+    let header = op.header();
+    let signature = op.signature().clone();
+    if let Some(entry) = op.entry() {
+        let entry_hashed = EntryHashed::with_pre_hashed(
+            entry.clone(),
+            header
+                .entry_hash()
+                .ok_or_else(|| DhtOpError::HeaderWithoutEntry(header.clone()))?
+                .clone(),
+        );
+        insert_entry(txn, entry_hashed)?;
+    }
+    let header_hashed = HeaderHashed::with_pre_hashed(header, op_light.header_hash().to_owned());
+    let header_hashed = SignedHeaderHashed::with_presigned(header_hashed, signature);
+    let op_order = OpOrder::new(op_light.get_type(), header_hashed.header().timestamp());
+    insert_header(txn, header_hashed)?;
+    insert_op_lite(txn, op_light, hash, is_authored, op_order)?;
+    Ok(())
+}
+
+/// Insert a [`DhtOpLight`] into the database.
+pub fn insert_op_lite(
+    txn: &mut Transaction,
+    op_lite: DhtOpLight,
+    hash: DhtOpHash,
+    is_authored: bool,
+    order: OpOrder,
+) -> StateMutationResult<()> {
+    let header_hash = op_lite.header_hash().clone();
+    let basis = op_lite.dht_basis().to_owned();
+    sql_insert!(txn, DhtOp, {
+        "hash": hash,
+        "type": op_lite.get_type(),
+        "basis_hash": basis,
+        "header_hash": header_hash,
+        "is_authored": is_authored,
+        "require_receipt": 0,
+        "blob": to_blob(op_lite)?,
+        "op_order": order,
+    })?;
```

Comment:
I mean... if you have the SQL and the execute in the same file, sure, it feels a little boiler-plate-y...

```rust
txn.execute(
    "INSERT INTO foo (bar) VALUES (:bar);",
    named_params! { ":bar": bar },
)?;
```

^^ I just had to write `bar` four times.

But if you split that out into separate files:

```sql
INSERT INTO foo (bar) VALUES (:bar);
```

```rust
const INSERT: &str = include_str!("sql");
txn.execute(INSERT, named_params! { ":bar": bar })?;
```

I still wrote `bar` four times, but I actually like how declarative it is... then you don't need the macros that separate you from the logic, adding to the cognitive overhead.

Reply:
I'm not so sure about this as all the inserts are identical and having to write the variable name 4 times for each field leaves opportunities for spelling mistakes.